azure/services/managedclusters/managedclusters.go

```diff
@@ -62,106 +61,58 @@ type ManagedClusterScope interface {
-func (s *Service) Reconcile(ctx context.Context) error {
-	ctx, _, done := tele.StartSpanWithLogger(ctx, "managedclusters.Service.Reconcile")
-	defer done()
-
-	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAKSServiceReconcileTimeout)
-	defer cancel()
-
-	managedClusterSpec := s.Scope.ManagedClusterSpec()
-	if managedClusterSpec == nil {
-		return nil
+func postCreateOrUpdateResourceHook(ctx context.Context, scope ManagedClusterScope, managedCluster *asocontainerservicev1.ManagedCluster, err error) error {
+	if err != nil {
+		return err
 	}
 
-	result, resultErr := s.CreateOrUpdateResource(ctx, managedClusterSpec, serviceName)
-	if resultErr == nil {
-		managedCluster, ok := result.(armcontainerservice.ManagedCluster)
-		if !ok {
-			return errors.Errorf("%T is not an armcontainerservice.ManagedCluster\n%v\n%v", result, result, managedCluster)
-		}
-		// Update control plane endpoint.
-		endpoint := clusterv1.APIEndpoint{
-			Host: ptr.Deref(managedCluster.Properties.Fqdn, ""),
-			Port: 443,
-		}
-		s.Scope.SetControlPlaneEndpoint(endpoint)
-
-		// Update kubeconfig data
-		// Always fetch credentials in case of rotation
-		adminKubeConfigData, userKubeConfigData, err := s.ReconcileKubeconfig(ctx, managedClusterSpec)
-		if err != nil {
-			return errors.Wrap(err, "error while reconciling adminKubeConfigData")
-		}
-
-		s.Scope.SetAdminKubeconfigData(adminKubeConfigData)
-		s.Scope.SetUserKubeconfigData(userKubeConfigData)
-
-		// This field gets populated by AKS when not set by the user. Persist AKS's value so for future diffs,
-		// the "before" reflects the correct value.
-		if id := managedCluster.Properties.IdentityProfile[kubeletIdentityKey]; id != nil && id.ResourceID != nil {
-			s.Scope.SetKubeletIdentity(*id.ResourceID)
-		}
-
-		s.Scope.SetOIDCIssuerProfileStatus(nil)
-		if managedCluster.Properties.OidcIssuerProfile != nil && managedCluster.Properties.OidcIssuerProfile.IssuerURL != nil {
-			s.Scope.SetOIDCIssuerProfileStatus(&infrav1.OIDCIssuerProfileStatus{
-				IssuerURL: managedCluster.Properties.OidcIssuerProfile.IssuerURL,
-			})
-		}
+	// Update control plane endpoint.
+	endpoint := clusterv1.APIEndpoint{
+		Host: ptr.Deref(managedCluster.Status.Fqdn, ""),
+		Port: 443,
 	}
-	s.Scope.UpdatePutStatus(infrav1.ManagedClusterRunningCondition, serviceName, resultErr)
-	return resultErr
-}
-
-// Delete deletes the managed cluster.
-func (s *Service) Delete(ctx context.Context) error {
-	ctx, _, done := tele.StartSpanWithLogger(ctx, "managedclusters.Service.Delete")
-	defer done()
+	scope.SetControlPlaneEndpoint(endpoint)
 
-	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
-	defer cancel()
-
-	managedClusterSpec := s.Scope.ManagedClusterSpec()
-	if managedClusterSpec == nil {
-		return nil
+	// Update kubeconfig data
+	// Always fetch credentials in case of rotation
+	adminKubeConfigData, userKubeConfigData, err := reconcileKubeconfig(ctx, scope, managedCluster.Namespace)
```

Comment:
@LochanRn If you could please test this PR to make sure it doesn't break your changes from #4008 I would very much appreciate it!

I tested the scenario where there is no AADProfile and where there is a managed AADProfile without DisableLocalAccounts and both of those seemed to produce valid kubeconfigs.

For the case where there is an AADProfile with DisableLocalAccounts, the secrets looked fine but if you could confirm that they actually work that would be great!

Reply:
Sure will test it and update here !

Reply:
@nojnhuh I was able to test disableLocalAccounts functionality and it is not breaking the change.